### PR TITLE
[dune] Add apidoc target using `odoc`

### PR DESCRIPTION
**Pull request by ejgallego**
We build the `@doc` target in the dune job.

- The documentation can be found in `_build/default/_doc/`
- We had to fix a couple of quoting problems.

Example documentation: https://coq.gitlab.io/-/coq/-/jobs/94498791/artifacts/_build/default/_doc/_html/coq/index.html

### CODE REVIEW

File: Makefile.dune
Comment (SkySkimmer) [inline Makefile.dune]: How well does this work when dune fails?

**Comment by SkySkimmer** [comment]
This has a lot of codeowners but they're mostly for the dependency on #6857

File: Makefile.dune
Comment (ejgallego) [reply Makefile.dune]: Yeah this should go away indeed.

File: tools/coq_dune.ml
Comment (gares) [inline tools/coq_dune.ml]: `Map.S` offers an `iter` operation directly

File: tools/coq_dune.ml
Comment (gares) [inline tools/coq_dune.ml]: fix the indentation

File: tools/coq_dune.ml
Comment (gares) [inline tools/coq_dune.ml]: ??

File: tools/coq_dune.ml
Comment (gares) [inline tools/coq_dune.ml]: you can write
```
let gen_sub n =
  let module List = struct let rec init = ... end in (* XXX : kill on OCaml >= 4.06 *)
  String.concat "/" (List.int ....)
```

File: tools/coq_dune.ml
Comment (gares) [inline tools/coq_dune.ml]: In the code below `let f x = match x with ... no use of x ...` should be replaced by `let f = function ...`

File: tools/coq_dune.ml
Comment (gares) [inline tools/coq_dune.ml]: Licence header is missing.

Also a line explaining what this utility is supposed to do would help.

The build instructions should be removed, they belong to the build system.

File: plugins/rtauto/plugin_base.dune
Comment (gares) [inline plugins/rtauto/plugin_base.dune]: There is something fishy here. All this code depends on coq. Coq is compiled with -rectypes and -rectypes is, AFAIK, viral. Why do you need to repeat it over and over? (Eg in findlib is the library using rectypes that declares it, not its users).

File: library/library.mld
Comment (gares) [inline library/library.mld]: what is CCI?

File: kernel/make_opcodes.sh
Comment (gares) [inline kernel/make_opcodes.sh]: same question as before

File: kernel/byterun/make_jumptbl.sh
Comment (gares) [inline kernel/byterun/make_jumptbl.sh]: does this file need to exist? can't you call sed from dune (as we do in Make)?

**Review by gares — DISMISSED** [review]

**Comment by gares** [comment]
I did review the entire diff, as usual stacked PR trick me

File: .gitlab-ci.yml
Comment (ejgallego) [reply .gitlab-ci.yml]: This is the package install solver develop by Debian people and used by opam to have better resolution of package constraints; it should go away once the new `odoc` version is released.

File: Makefile.dune
Comment (ejgallego) [reply Makefile.dune]: Noted.

File: Makefile.dune
Comment (ejgallego) [reply Makefile.dune]: Noted, thanks.

File: config/dune
Comment (ejgallego) [reply config/dune]: Will likely need more tweaks indeed.

File: coq.opam
Comment (ejgallego) [reply coq.opam]: No need to depend on `make`, it is likely that in the future once we remove this `vodeps` step the build target is empty as OPAM will know how to build Dune packages automatically.

File: dev/ci/ci-pidetop.sh
Comment (ejgallego) [reply dev/ci/ci-pidetop.sh]: This has to be handled upstream, `jbuilder` and `dune` are not compatible as they use different languages.

File: dev/doc/build-system.dev.txt
Comment (ejgallego) [reply dev/doc/build-system.dev.txt]: Updated, thanks.

File: dev/doc/build-system.dev.txt
Comment (ejgallego) [reply dev/doc/build-system.dev.txt]: Umm, I've reworded that part but I have to disagree with this comment, this is not about selling, but about indicating that noting out of the ordinary is happening here and IMO these two things are accurate descriptions of what is going on.

File: dev/doc/build-system.txt
Comment (ejgallego) [reply dev/doc/build-system.txt]: I mean that you don't need to call `configure`.

Comment (ejgallego) [reply dev/doc/build-system.txt]: I am indeed anticipating here a patch that basically eliminates configure for Dune-based builds [and for Makefile if so desired].

File: dev/doc/build-system.txt
Comment (ejgallego) [reply dev/doc/build-system.txt]: What are you talking about? I am sorry I do not follow. This is, as noted in the proper PR, still a TODO. Selling what? pffff

File: tools/coq_dune.ml
Comment (ejgallego) [reply tools/coq_dune.ml]: I will, but honestly I find this much easier to read.

File: tools/coq_dune.ml
Comment (ejgallego) [reply tools/coq_dune.ml]: Indeed, thanks!

File: tools/coq_dune.ml
Comment (ejgallego) [reply tools/coq_dune.ml]: ??

File: tools/coq_dune.ml
Comment (ejgallego) [reply tools/coq_dune.ml]: Header added, note added, bootstrap instructions kept as they important, note added.

File: tools/coq_dune.ml
Comment (ejgallego) [reply tools/coq_dune.ml]: I guess this is a matter of style and we don't enforce it in Coq, right?

File: plugins/rtauto/plugin_base.dune
Comment (ejgallego) [reply plugins/rtauto/plugin_base.dune]: Umm, IIRC my projects using findlib and `META.coq` needed to add `-rectypes` manually; maybe I did something wrong. It is easy to try and see.

Anyways, Dune now supports a per-project `flags` declaration, so these repetition will go away.

File: library/library.mld
Comment (ejgallego) [reply library/library.mld]: This was in the old text, I dunno.

File: kernel/byterun/make_jumptbl.sh
Comment (ejgallego) [reply kernel/byterun/make_jumptbl.sh]: I think it should be the other way around and Make should call this file instead of inlining code, so we share the definition.

I could indeed inline the code tho.

**Comment by ejgallego** [comment]
> I did review the entire diff, as usual stacked PR trick me.

Thanks for the review, comments addressed. Sorry for the stacked PR thing, I dunno, I was asked to split some parts off so again the tradeoff is hard. Note that a few concerns are being due to the original PR not being yet ready, in particular regarding documentation.

Some other stuff such as better configure will wait for a post-merge I think.

**Comment by ejgallego** [comment]
> Some other stuff such as better configure will wait for a post-merge I think.

Note that many of your comments are addressed in the proper PR including plans before merge.

File: tools/coq_dune.ml
Comment (gares) [reply tools/coq_dune.ml]: I do

File: library/library.mld
Comment (gares) [reply library/library.mld]: ok, then it is the french way of spelling CIC. Please fix

File: kernel/byterun/make_jumptbl.sh
Comment (gares) [reply kernel/byterun/make_jumptbl.sh]: This is a good idea. Please call the two .sh scripts from Make, so that there is no chance of getting the two build systems out of sync.

File: config/dune
Comment (gares) [reply config/dune]: you can pass `-profile devel -extra-stuff` if you need to tweak more. `-local` and `-warn-error` are implied by `-profile devel`.

File: dev/doc/build-system.txt
Comment (gares) [reply dev/doc/build-system.txt]: I think a comment of mine got lost, I'll post it now on the .dev file

File: tools/coq_dune.ml
Comment (ejgallego) [reply tools/coq_dune.ml]: > I do

What does this mean?

File: dev/doc/build-system.txt
Comment (ejgallego) [reply dev/doc/build-system.txt]: > for example I don't see what is the difference with make.

Well the difference is pretty obvious as Make knows nothing about OCaml libraries and thus we had to roll our own library management system that has quite a bit of issues [documented in the PR].

I will make this more explicit tho.

Comment (ejgallego) [reply dev/doc/build-system.txt]: A good example is that developers now have to mess with `.mlllib` files, with Dune this is gone.

File: coq.opam
Comment (rgrinberg) [reply coq.opam]: Why can't you use the default `@install` target here. Also, building for opam without `-p` is not recommended. `-p` will disable warnings and do other good things that are necessary for a release build.

Comment (ejgallego) [reply coq.opam]: > Why can't you use the default @install target here. Also, building for opam without -p is not recommended. -p will disable warnings and do other good things that are necessary for a release build.

@rgrinberg We do, note that this is the preliminary step to generate `dune` files for the Coq libraries; I ignore how to make this work in a single call due to new `dune` files being added.

File: tools/coq_dune.ml
Comment (gares) [reply tools/coq_dune.ml]: I gave you a reason that is not aestetic. And it has nothing to do with Coq.
You are free to think it is aestetic, but this is not what I think. I'm not bykeshedding your code.

Hence (I do) care about this matter and I'm asking to use `function` when the input is use linearly and immediately destructed. `function` plays this role in OCaml.

File: tools/coq_dune.ml
Comment (gares) [reply tools/coq_dune.ml]: I don't have a strong opinion on the solution I propose, but having code in comments that is supposed to be uncommented later on is a bad idea: the actual code could change, and nobody is going to fix the comment, so you may end up uncommenting old code.
What I wanted to achieve is code that stays the same.
You can also do the following if you prefer.
```
module List = struct include List let rec init .... end (* XXX remove on OCaml 4.06 *)
```

File: plugins/rtauto/plugin_base.dune
Comment (gares) [reply plugins/rtauto/plugin_base.dune]: If dune supports this option, why don't you use it? What do you mean by "will go away"?

Comment (ejgallego) [reply plugins/rtauto/plugin_base.dune]: > If dune supports this option, why don't you use it?

As noted in the original PR TODO list this was pending on the release of Dune 1.0 so will be implemented soon and before merging.

File: tools/coq_dune.ml
Comment (ejgallego) [reply tools/coq_dune.ml]: Will this trigger a warning when compiled with more recent OCaml?

Comment (ejgallego) [reply tools/coq_dune.ml]: We have plenty of code in that situation sadly but it is not an easy problem to solve, other than duplicating many modules.

Comment (gares) [reply tools/coq_dune.ml]: If the problem is a warning you can suppress it using an attribute.
I don't see why it is hard to solve in this case.

**Comment by ejgallego** [comment]
Ready to undergo review.

File: dev/ci/README.md
Comment (ejgallego) [reply dev/ci/README.md]: Updated.

Comment (ejgallego) [reply dev/ci/README.md]: The docs should be installed by the way, I'll fix that too.

File: kernel/kernel.mld
Comment (Zimmi48) [inline kernel/kernel.mld]: I didn't see this anywhere in the odoc rendered HTML.

Comment (ejgallego) [reply kernel/kernel.mld]: This is placed on the `coq/kernel.html` file, but it is not linked from anywhere. I think we will have to wait for `odoc` 1.3.0 [or maybe it will be called 2.0] to see what we do; there are many new features being added so for now really this PR has mostly "witness" status so we can see how odoc does and we are sure docs build with it.

**Comment by ejgallego** [comment]
opam deps moved to docker.

**Comment by ejgallego** [comment]
I have make the documentation build to depend on the dune build, however the doc building is so fast that I am not sure it does make sense.

**Comment by ejgallego** [comment]
@SkySkimmer suggest to include the changes to Docker in this PR.

**Comment by Zimmi48** [comment]
>however the doc building is so fast that I am not sure it does make sense

probably not indeed

**Review by SkySkimmer — CHANGES_REQUESTED** [review]

File: library/library.mld
Comment (ejgallego) [reply library/library.mld]: I may remove the mld files at all I think, until we have a better plan for the API docs.

Comment (ejgallego) [reply library/library.mld]: They were just there to illustrate what you can do, but new odoc will open up some more possibilities so we can wait a bit.

**Comment by ejgallego** [comment]
Fixed.

**Review by SkySkimmer — APPROVED** [review]

**Pull request merged**
